### PR TITLE
Create data categorizer

### PR DESCRIPTION
**Pull request by ben-heil**
This PR adds a script to the repository that is used for separating samples from the SRA into case and control categories. The labeled samples can then be used for supervised learning.

### CODE REVIEW

File: find_studies.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+            print('Parsing XML...')
+            out = open('test.txt', 'w')
+            out.write(strip_unicode(result.text))
+            root = ET.fromstring(strip_unicode(result.text))
+
+            return root
+        except requests.exceptions.HTTPError:
+            attempts += 1
+            print('Downloading dataset failed {} times'.format(attempts))
+            sleep(1)
```
Comment (dhimmel) [inline find_studies.py]: would suggest some sort of exponential backoff where the sleep increases as attempts grows.

File: find_studies.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+    -----
+    The tree will be saved to the disk as <out_file>.xml, and the rules will be saved
+    as <out_file>_rules.json
+
+    '''
+    tree = ET.ElementTree(root)
+    tree.write(results_out)
+
+    #rules_json = json.dumps(rules)
+    rules_file = open(rules_out, 'w')
```
Comment (dhimmel) [inline find_studies.py]: should be using context manager here to close file when json.dump is complete

```
with open() as rules_file:
   dump
```

File: find_studies.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+        print('in conditional')
+        tree = ET.parse(args.previous_xml)
+        root = tree.getroot()
+
+        studies = root.findall('.//EXPERIMENT_PACKAGE_SET')
+        for study in studies:
+            previously_seen_studies.add(study.get('study_id'))
+    if args.previous_rules is not None:
+        rule_file = open(args.previous_rules)
+        rule_dict = json.load(rule_file)
```
Comment (dhimmel) [inline find_studies.py]: use context manager

File: find_studies.py [hunk 4]
```diff
@@ -0,0 +1,457 @@
...
+        print(rule_dict)
+        rule_file.close()
+
+    data = None
+    with urllib.request.urlopen('http://metasra.biostat.wisc.edu/api/v01/'
+        'samples.json?and=UBERON:0000178&sampletype=tissue') as url:
+        data = json.loads(url.read().decode())
```
Comment (dhimmel) [inline find_studies.py]: requests is imported so I'd just use `requests.get().json()`

**Review by dhimmel — COMMENTED** [review]
Hard to review without seeing what the data (or a subset of it) looks like. Would also help to see the command output (even just as a comment on the PR).

The printed messages should probably go to stderr rather than stdout IMO. Possibly should use `logging` module. Depends perhaps on whether this is an important script you plan to use a lot, or just something you need to work once.

**Comment by ben-heil** [comment]
The initial JSON data the script retrieves from MetaSRA are in the format listed on [this page](http://metasra.biostat.wisc.edu/api.html)


The input from SRA is in xml; the parts that I use are of the format
```xml
  <EXPERIMENT_PACKAGE>
    <STUDY center_name="dbGaP" alias="phs000464" accession="SRP011999"></STUDY>
    <SAMPLE alias="TARGET-15-SJMPAL043510-09A-01R" accession="SRS2623322">
      <TITLE> Example </TITLE>
    </SAMPLE>
  </EXPERIMENT_PACKAGE>
 ```
The outputs are of the same format, but are composed of multiple EXPERIMENT_PACKAGE elements wrapped in an EXPERIMENT_PACKAGE_SET.

The rules are output as a json object mapping study ids to lists of rules. An example rules file is:
```json
{"SRP011999": [["case", "test"]], "SRP012682": [["control", "test2"]], "SRP012000": [["case", "tumor"], ["control", "non-tumor"], ["invalid", "dna"]], "SRP108709": []}
```

File: find_studies.py (hunk 1 above)
Comment (ben-heil) [reply find_studies.py]: Increasing the amount of time sleeping is a good idea, I added a linear backoff though

**Comment by ben-heil** [comment]
I only intend to run this script a few times so I went with writing to stderr instead of using the logging utility

**Review by dhimmel — APPROVED** [review]

**Pull request merged**